### ISSUE TITLE
Option Proposal: `no-undef` covers inside of `typeof` operators

### ISSUE THREAD

**Issue by mysticatea** [eslint/eslint#3684]
Some days ago, I mistook modifying references which are inside of `typeof` operators while I refactored my code.

So I will request an option to catch references which are inside of `typeof` operators.
I thought the effect of the exception which ignores references which are inside of `typeof` operators is limited, because I guess that operator aims to explicit/implicit global variables.

```js
{
    "no-undef": [2, {"typeof": true}]
}
```

The following pattern are considered errors by an addition.

```js
if (typeof anUndefinedVar === "string") {
}
```

**Comment by michaelficarra** [comment]
I don't understand why you would want that to be considered an error. It's always safe to do that.

**Comment by mysticatea** [comment]
@michaelficarra It's a difference between "valid in the language spec" and "valid in the application logic".

For example:

```js
function foo(a, b) {
    if (typeof a === "function") {
    }
}

// rename a -> z

function foo(z, b) {
    if (typeof a === "function") {
    }
}
```

This code is not reported by eslint, but it's a mistake.
I want an option that pattern is reported.

**Comment by michaelficarra** [comment]
Then how would you check the existence of a variable?

**Comment by mysticatea** [comment]
To skip here if the option is specified: https://github.com/eslint/eslint/blob/master/lib/rules/no-undef.js#L81-L83

**Comment by michaelficarra** [comment]
No, in your program. This is a useful pattern. What would you replace it with?

**Comment by nzakas** [comment]
I think this is fine as an option that is off by default.

**Comment by mysticatea** [comment]
@michaelficarra I don't need it in my app. If I needed, I would define the global veriable in my .eslintrc. If I don't define, I cannot use references of the veriable outside of typeof expr.

@nzakas Yes, off by default. Thanks!
